File: backend/src/services/excelService.js
'use strict';
const ExcelJS = require('exceljs');

/**
 * Generate an Excel workbook from data.
 * columns: [{ header, key, width }]
 * rows: array of objects
 */
async function generateXlsx(columns, rows, sheetName = 'Dane') {
  const workbook = new ExcelJS.Workbook();
  workbook.creator = 'System ZUP';
  workbook.created = new Date();

  const sheet = workbook.addWorksheet(sheetName);
  sheet.columns = columns.map(c => ({
    header: c.header,
    key: c.key,
    width: c.width || 20,
  }));

  // Style header row
  const headerRow = sheet.getRow(1);
  headerRow.font = { bold: true, color: { argb: 'FFFFFFFF' } };
  headerRow.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF2563EB' } };
  headerRow.alignment = { vertical: 'middle', horizontal: 'center' };
  headerRow.height = 22;

  // Add data
  for (const row of rows) {
    sheet.addRow(row);
  }

  // Borders
  sheet.eachRow((row) => {
    row.eachCell((cell) => {
      cell.border = {
        top: { style: 'thin' },
        left: { style: 'thin' },
        bottom: { style: 'thin' },
        right: { style: 'thin' },
      };
    });
  });

  return workbook;
}

/**
 * Send xlsx as HTTP response.
 */
async function sendXlsxResponse(res, columns, rows, filename, sheetName) {
  const workbook = await generateXlsx(columns, rows, sheetName);
  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
  res.setHeader('Content-Disposition', `attachment; filename="${filename}.xlsx"`);
  await workbook.xlsx.write(res);
  res.end();
}

/**
 * Send CSV as HTTP response.
 */
function sendCsvResponse(res, columns, rows, filename) {
  res.setHeader('Content-Type', 'text/csv; charset=utf-8');
  res.setHeader('Content-Disposition', `attachment; filename="${filename}.csv"`);
  res.write('\uFEFF'); // BOM for Excel
  const headers = columns.map(c => `"${c.header}"`).join(';');
  res.write(headers + '\n');
  for (const row of rows) {
    const line = columns.map(c => {
      const val = row[c.key];
      if (val === null || val === undefined) return '""';
      return `"${String(val).replace(/"/g, '""')}"`;
    }).join(';');
    res.write(line + '\n');
  }
  res.end();
}

module.exports = { generateXlsx, sendXlsxResponse, sendCsvResponse };

File: backend-py/services/excel_service.py
from __future__ import annotations
import csv
import io
from typing import Any

from fastapi.responses import StreamingResponse


def send_xlsx(data: list[dict], cols: list[dict], filename: str, sheet_name: str = "Sheet1") -> StreamingResponse:
    try:
        import openpyxl
        from openpyxl.styles import Font, PatternFill, Alignment
    except ImportError:
        raise ImportError("openpyxl is not installed")

    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = sheet_name

    header_fill = PatternFill(start_color="366092", end_color="366092", fill_type="solid")
    header_font = Font(color="FFFFFF", bold=True)

    for i, col in enumerate(cols, 1):
        cell = ws.cell(row=1, column=i, value=col["header"])
        cell.fill = header_fill
        cell.font = header_font
        cell.alignment = Alignment(horizontal="center")

    for row_idx, row in enumerate(data, 2):
        for col_idx, col in enumerate(cols, 1):
            value = row.get(col["key"])
            if value is not None and hasattr(value, "isoformat"):
                value = str(value)
            ws.cell(row=row_idx, column=col_idx, value=value)

    for col in ws.column_dimensions:
        ws.column_dimensions[col].width = 18

    output = io.BytesIO()
    wb.save(output)
    output.seek(0)

    return StreamingResponse(
        output,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename={filename}.xlsx"},
    )


def send_csv(data: list[dict], cols: list[dict], filename: str) -> StreamingResponse:
    output = io.StringIO()
    writer = csv.writer(output, delimiter=";")
    writer.writerow([col["header"] for col in cols])
    for row in data:
        writer.writerow([row.get(col["key"], "") for col in cols])
    output.seek(0)

    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": f"attachment; filename={filename}.csv"},
    )
